File: packages/shared-types/src/organization.ts
import type { FolderKind, TagScope } from "./enums"
import type { ArticleId, FeedId, FolderId, TagId } from "./ids"
import type { HexColor, ISODateTimeString, Nullable } from "./primitives"

export interface FolderDto {
  id: FolderId
  name: string
  parentId: Nullable<FolderId>
  sortOrder: number
  kind: FolderKind
}

export interface TagDto {
  id: TagId
  name: string
  scope: TagScope
  color: Nullable<HexColor>
  createdAt: ISODateTimeString
}

export interface FeedTagDto {
  feedId: FeedId
  tagId: TagId
}

export interface ArticleTagDto {
  articleId: ArticleId
  tagId: TagId
}

File: crates/core-domain/src/model/organization.rs
use super::{ArticleId, FeedId, FolderId, FolderKind, HexColor, IsoDateTime, TagId, TagScope};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Folder {
    pub id: FolderId,
    pub name: String,
    pub parent_id: Option<FolderId>,
    pub sort_order: i64,
    pub kind: FolderKind,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Tag {
    pub id: TagId,
    pub name: String,
    pub scope: TagScope,
    pub color: Option<HexColor>,
    pub created_at: IsoDateTime,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FeedTag {
    pub feed_id: FeedId,
    pub tag_id: TagId,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ArticleTag {
    pub article_id: ArticleId,
    pub tag_id: TagId,
}
